zsl/resource/json_server_resource.py

```diff
@@ -0,0 +1,272 @@
+"""
+:mod:`zsl.resource.json_server_resource`
+----------------------------------------
+"""
+from __future__ import (absolute_import, division,
+                        print_function, unicode_literals)
+from builtins import *
+
+import logging
+import math
+import re
+
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy import or_
+
+from .model_resource import ModelResource
+from zsl.interface.resource import ResourceResult
+from zsl.service.service import transactional
+from zsl.resource.resource_helper import filter_from_url_arg, flat_model, model_tree
+
+import http.client
+from flask import request
+
+
+def status_code(status):
+    """Solve Py2/3 problem."""
+    if hasattr(status, 'value'):
+        return status.value
+    else:
+        return status
+
+NOT_FOUND = ResourceResult(
+    body={},
+    status=status_code(http.client.NOT_FOUND)
+)
+
+NOT_IMPLEMENTED = ResourceResult(
+    body={},
+    status=status_code(http.client.NOT_IMPLEMENTED)
+)
+
+# any other arguments from these are considered as `property_name(_operator)=some_vaule` filter
+SKIPPED_ARGUMENTS = set(['callback', '_', 'q', '_start', '_end', '_sort', '_order', '_limit', '_embed', '_expand'])
+
+# first group is the column name, then it can have a . separator or an operator suffix
+re_column_name = re.compile(r'^([^.]*?)(\..*?)?(_lte|_gte|_ne|_like)?$')
+
+
+def page_arg(p):
+    # type: (int) -> str
+    """Create a page argument from int."""
+    return 'page=' + str(p)
+
+
+def get_link_pages(page, per_page, count, page_url):
+    # type: (int, int, int, str) -> Dict[str, str]
+    """Create link header for page metadata.
+    
+    :param page: current page
+    :param per_page: page limit
+    :param count: count of all resources
+    :param page_url: url for resources
+    :return: dictionary with name of the link as key and its url as value
+    """
+    current_page = page_arg(page)
+    links = {}
+    end = page * per_page
+
+    if page > 1:
+        links['prev'] = page_url.replace(current_page, page_arg(page-1))
+
+    if end < count:
+        links['next'] = page_url.replace(current_page, page_arg(page+1))
+
+    if per_page < count:
+        links['first'] = page_url.replace(current_page, page_arg(1))
+        links['last'] = page_url.replace(current_page, page_arg(math.ceil(count / per_page)))
```

Comment:
i'd prefer (count + per_page - 1) // per_page computation
http://stackoverflow.com/questions/14822184/is-there-a-ceiling-equivalent-of-operator-in-python
plus if per_page and count are insanely large the above computation may fail